_posts/2021-01-18-smtchecker-and-synthesis-of-external-functions.md

```diff
@@ -0,0 +1,420 @@
+		x = _x;
+	}
+
+	function run() mutex public {
+		uint xPre = x;
+		unknown.run();
+		assert(xPre == x);
+	}
+}
+```
+
+[Spacer](https://spacer.bitbucket.io/),
+the Horn solver used by the CHC engine, automatically infers the inductive
+invariant `lockPre => lockPost && lockPre => (xPost = xPre)` for that specific
+external call, that is, it learns that
+
+- if `lock = true` before the call, it must also be `true` after the call.
+- if `lock = true` before the call, `x` cannot be modified by reentrant calls.
+
+From the local context it knows that `lock = true` before the call.  Therefore,
+it must be true that **no reentrant call** can modify `x`, and the property is
+correct.
+
+Another important feature is the one that names this post. In case *it is*
+possible to violate the assertion, the solver synthesizes the behavior of the
+externally called unknown code, telling us exactly what needs to be done to
+break the property. If `Mutex.run` is not guarded by `lock`, the SMTChecker
+reports
+
+```bash
+Warning: CHC: Assertion violation happens here.
+Counterexample:
+x = 1, lock = false, unknown = 1
+
+Transaction trace:
+Mutex.constructor(1)
+State: x = 0, lock = false, unknown = 1
+Mutex.run()
+    unknown.run() -- untrusted external call, synthesized as:
+        Mutex.set(1) -- reentrant call
+  --> mutex.sol:34:3:
+   |
+34 | 		assert(xPre == x);
+   | 		^^^^^^^^^^^^^^^^^
+```
+
+Here the solver finds out that all the unknown code needs to do is call
+`Mutex.set` once, and the property will be violated at the end of `Mutex.run`.
+
+Note that `unknown = 1` means the address of the `Unknown` contract called by
+`Mutex`, where the actual address of the deployed contract doesn't really
+matter, as long as it implements a function `run()` that calls `Mutex.set(1)`.
+
+---
+
+The following sample contract shows the counterexample report in case the external
+unknown function needs to make multiple calls:
+
+```solidity
+//SPDX-License-Identifier: GPL-v3
+pragma solidity ^0.8.0;
+pragma experimental SMTChecker;
+
+interface Unknown {
+	function run() external;
+}
+
+contract Test {
+	uint x;
+	uint y;
+
+	Unknown immutable unknown;
+
+	constructor(Unknown _u) {
+		require(address(_u) != address(0));
+		unknown = _u;
+	}
+
+	function incX() public { ++x; }
+	function incY() public { ++y; }
+
+	function f() public {
+		unknown.run();
+		assert(x < 2 || y < 2);
+	}
+}
+```
+
+The property is false, and the SMTChecker reports:
+
+```bash
+$ solc external_inc.sol --model-checker-engine chc
+
+Warning: CHC: Assertion violation happens here.
+Counterexample:
+x = 3, y = 1, unknown = 1
+
+Transaction trace:
+Test.constructor(1)
+State: x = 0, y = 0, unknown = 1
+Test.f()
+    unknown.run() -- untrusted external call, synthesized as:
+        Test.incX() -- reentrant call
+        Test.incX() -- reentrant call
+        Test.incX() -- reentrant call
+        Test.incY() -- reentrant call
+  --> external_inc.sol:25:3:
+   |
+25 | 		assert(x < 3 || y < 1);
```

Comment:
This is different that the property in the exampe code above

Reply:
Indeed!